### PR TITLE
CRM-19773 - Call hook_civicrm_selectWhereClause when checking event permissions

### PR DESCRIPTION
**Pull request by colemanw**
* [CRM-19773: Call hook_civicrm_selectWhereClause from the BAOs](https://issues.civicrm.org/jira/browse/CRM-19773)

### CODE REVIEW

File: CRM/Event/BAO/Event.php
Comment (nganivet) [inline CRM/Event/BAO/Event.php]: So cool ...

File: CRM/Event/BAO/Event.php
Comment (nganivet) [inline CRM/Event/BAO/Event.php]: Wouldn't keeping $allEvents = CRM_Event_PseudoConstant::event(NULL, TRUE); be more efficient than an API call?

File: CRM/Event/Page/ManageEvent.php
Comment (nganivet) [inline CRM/Event/Page/ManageEvent.php]: So if I understand correctly we do not need to filter in the Manage Events page since the filtering takes place in CRM_Event_BAO_Event::checkPermissions() ?

Comment (colemanw) [inline CRM/Event/Page/ManageEvent.php]: Right.

File: CRM/Event/BAO/Event.php
Comment (colemanw) [inline CRM/Event/BAO/Event.php]: The api call invokes hook_civicrm_selectWhereClause and the pseudoconstant lookup does not (I did add it there, but because of the `$all` flag being set to `true` it skips the hook). The pseudoconstant function is deprecated anyway.

File: CRM/Event/BAO/Event.php
Comment (colemanw) [inline CRM/Event/BAO/Event.php]: Yep, that's 3 levels of joins to get the state_province name :)

**Comment by colemanw** [comment]
@civicrm-builder retest this please

**Comment by colemanw** [comment]
@civicrm-builder retest this please

**Comment by eileenmcnaughton** [comment]
test this please

**Comment by eileenmcnaughton** [comment]
It would be good to get a test on this

**Review by jitendrapurohit — APPROVED** [review]
Tested, works fine. Also added a test https://github.com/civicrm/civicrm-core/pull/10125. Not sure why the build isn't passing here.

**Comment by colemanw** [comment]
@civicrm-builder test this please

File: CRM/Event/BAO/Event.php
Comment (jitendrapurohit) [inline CRM/Event/BAO/Event.php]: Build shows a break at this point - maybe all failing test needs `$this->createLoggedInUser();` in them to validate `'created_id' => 'user_contact_id'` param?

**Comment by eileenmcnaughton** [comment]
I don't know why the test is failing but once you get it to pass it's good to merge based on @jitendrapurohit's review & my reading of the code - adding merge-on-pass

**Comment by eileenmcnaughton** [comment]
@colemanw this has been reviewed AND @jitendrapurohit added a unit test! - but can't merge as tests are failing

I think the trick is to allow 'user_contact_id' to be converted to NULL & rely on whether it is mandatory or not to enforce the id - in the case of the failing tests it is not

```
  if (!empty($fieldValue) || $fieldValue === '0' || $fieldValue === 0) {
    // if value = 'user_contact_id' (or similar), replace value with contact id
    if (!is_numeric($fieldValue) && is_scalar($fieldValue)) {
      $realContactId = _civicrm_api3_resolve_contactID($fieldValue);
      if ('unknown-user' === $realContactId) {
        throw new API_Exception("\"$fieldName\" \"{$fieldValue}\" cannot be resolved to a contact ID", 2002, array('error_field' => $fieldName, "type" => "integer"));
      }
      elseif (is_numeric($realContactId)) {
        $fieldValue = $realContactId;
      }
    }
    if (!empty($fieldInfo['pseudoconstant']) || !empty($fieldInfo['options'])) {
      _civicrm_api3_api_match_pseudoconstant($fieldValue, $entity, $fieldName, $fieldInfo, $op);
    }

    // After swapping options, ensure we have an integer(s)
    foreach ((array) ($fieldValue) as $value) {
      if ($value && !is_numeric($value) && $value !== 'null' && !is_array($value)) {
        throw new API_Exception("$fieldName is not a valid integer", 2001, array('error_field' => $fieldName, "type" => "integer"));
      }
    }
```

**Comment by eileenmcnaughton** [comment]
from function _civicrm_api3_validate_integer(&$params, $fieldName, &$fieldInfo, $entity) {

**Comment by eileenmcnaughton** [comment]
@colemanw needs rebasing but will hopefully pass now

**Comment by eileenmcnaughton** [comment]
@colemanw this appears to have caused a regression due to default limit of 25 on the api calls

**Comment by mfb** [comment]
Yes, it's fixed if I add `'options' => array('limit' => 0)` to these API calls.  See also https://civicrm.stackexchange.com/questions/18516/events-missing-after-upgrade-to-4-7-19

**Comment by colemanw** [comment]
I've opened a PR - #10327

**Pull request merged**
